### ISSUE TITLE
Unexpected `regexp/no-useless-escape` report with `RegExp` constructor

### ISSUE THREAD

**Issue by JounQin** [ota-meshi/eslint-plugin-regexp#701]
**Information:**
- ESLint version: v8.48.0
- `eslint-plugin-regexp` version: v2.2.0

**Description**

```js
new RegExp('[\\.]')
```

It's definitely different with `new RegExp('[\.]')`.

**Comment by RunDevelopment** [comment]
On the playground, the auto fix wants to simplify it to `new RegExp('[.]')`, which is correct.

![image](https://github.com/ota-meshi/eslint-plugin-regexp/assets/20878432/ac6e939f-b3e0-43e9-bacf-993456e1d37c)

I think the cause of the confusing is that the error message says: `Unnecessary escape character: \..`. The message only mentions one backslash because there is only one backslash in contained in the source of the regex. We just need to write one backslash as `\\` because of JS string syntax.

In general, we should probably do some kind of source mapping here. E.g. in this case, I think the error message should have been something like: `Unnecessary escape character '\.' (written as '\\.' in source code) can be replaced with '.'.`

**Comment by JounQin** [comment]
@RunDevelopment The worst thing is when using something like

```js
var regexpStr = '\\.[\\.]'
var regexp = new RegExp(regexpStr)
```

the error is not reported at `regexpStr` but `regexp` without source location info, so that's very confusing.

**Comment by RunDevelopment** [comment]
> the error is not reported at `regexpStr` but `regexp` without source location info, so that's very confusing.

![image](https://github.com/ota-meshi/eslint-plugin-regexp/assets/20878432/5d6b78b4-7dd5-4a14-a195-df136cccbb4f)

As you can see, the error is reported on `regexpStr` as expected. I think you meant a case like this:

![image](https://github.com/ota-meshi/eslint-plugin-regexp/assets/20878432/06d7e1b7-928d-45ed-a76e-404da8268ac1)

Cases where a single string is used by >1 regexes are hard, because errors might not be the same across all regexes. Example:

![image](https://github.com/ota-meshi/eslint-plugin-regexp/assets/20878432/7b174ac1-64a4-4930-95d4-30fc20c29041)

The regex with the `i` flag wants to simplify `[Aab]` to `[ab]`, but this isn't valid because of the regex without the `i` flag.

In general, cases where a string is read by something other than a single `RegExp` constructor are difficult to deal with, because there might be good reasons to not change the string.
